include/nbl/builtin/hlsl/sort/counting.hlsl

```diff
@@ -0,0 +1,138 @@
+#define _NBL_BUILTIN_HLSL_SORT_COUNTING_INCLUDED_
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace sort
+{
+
+template<
+    uint16_t GroupSize,
+    uint16_t KeyBucketCount,
+    typename Key,
+    typename KeyAccessor,
+    typename ValueAccessor,
+    typename HistogramAccessor,
+    typename SharedAccessor,
+    bool robust=false
+>
+struct counting
+{
+    uint32_t inclusive_scan(uint32_t value, NBL_REF_ARG(SharedAccessor) sdata)
+    {
+        return workgroup::inclusive_scan < plus < uint32_t >, GroupSize >::
+                template __call <SharedAccessor>(value, sdata);
+    }
+
+    uint32_t toroidal_histogram_add(uint32_t tid, uint32_t sum, NBL_REF_ARG(HistogramAccessor) histogram, NBL_REF_ARG(SharedAccessor) sdata, const CountingParameters<Key> params)
+    {
+        sdata.workgroupExecutionAndMemoryBarrier();
+
+        sdata.set(tid % GroupSize, sum);
+        uint32_t shifted_tid = (tid + glsl::gl_SubgroupSize() * params.workGroupIndex) % GroupSize;
+
+        sdata.workgroupExecutionAndMemoryBarrier();
+
+        return histogram.atomicAdd((tid / GroupSize) * GroupSize + shifted_tid, sdata.get(shifted_tid));
+    }
+
+    void build_histogram(NBL_REF_ARG( KeyAccessor) key, NBL_REF_ARG(SharedAccessor) sdata, const CountingParameters<Key> params)
+    {
+        uint32_t tid = workgroup::SubgroupContiguousIndex();
+
+        for (; tid < KeyBucketCount; tid += GroupSize) {
+            sdata.set(tid, 0);
+        }
+
+        sdata.workgroupExecutionAndMemoryBarrier();
+
+        uint32_t index = params.workGroupIndex * GroupSize * params.elementsPerWT + tid % GroupSize;
+        uint32_t endIndex = min(params.dataElementCount, index + GroupSize * params.elementsPerWT);
+
+        for (; index < endIndex; index += GroupSize)
+        {
+            uint32_t k = key.get(index);
+            if (robust && (k<params.minimum || k>params.maximum) )
+                continue;
+            sdata.atomicAdd(k - params.minimum, (uint32_t) 1);
+        }
+
+        sdata.workgroupExecutionAndMemoryBarrier();
+    }
+
+    void histogram(NBL_REF_ARG( KeyAccessor) key, NBL_REF_ARG(HistogramAccessor) histogram, NBL_REF_ARG(SharedAccessor) sdata, const CountingParameters<Key> params)
+    {
+        build_histogram(key, sdata, params);
+
+        uint32_t tid = workgroup::SubgroupContiguousIndex();
+        uint32_t histogram_value = sdata.get(tid);
+
+        sdata.workgroupExecutionAndMemoryBarrier();
+
+        uint32_t sum = inclusive_scan(histogram_value, sdata);
+        toroidal_histogram_add(tid, sum, histogram, sdata, params);
+
+        const bool is_last_wg_invocation = tid == (GroupSize - 1);
+        const uint16_t adjusted_key_bucket_count = ((KeyBucketCount - 1) / GroupSize + 1) * GroupSize;
+
+        for (tid += GroupSize; tid < adjusted_key_bucket_count; tid += GroupSize)
+        {
+            if (is_last_wg_invocation)
+            {
+                uint32_t startIndex = tid - tid % GroupSize;
```

Comment:
I don't get why you calculate the index with `%`

Anyway if you go back to a uniform loop, you can compute it as `i*GroupSize`

Reply:
Done